app/scripts/metamask-controller.js

```diff
@@ -6612,22 +6792,281 @@ export default class MetamaskController extends EventEmitter {
+      if (
+        ![
+          MESSAGE_TYPE.WALLET_CREATE_SESSION,
+          MESSAGE_TYPE.WALLET_INVOKE_METHOD,
+          MESSAGE_TYPE.WALLET_GET_SESSION,
+          MESSAGE_TYPE.WALLET_REVOKE_SESSION,
+        ].includes(req.method)
+      ) {
+        return end(new Error('Invalid method')); // TODO: Use a proper error
+      }
+      return next();
+    });
+
+    // TODO: Uncomment this when wallet lifecycle methods are added to api-specs
+    engine.push(multichainMethodCallValidatorMiddleware);
+    const middlewareMaker = makeMethodMiddlewareMaker([
+      walletRevokeSession,
+      walletGetSession,
+      walletInvokeMethod,
+      walletCreateSession,
+    ]);
+
+    engine.push(
+      middlewareMaker({
+        grantPermissions: this.permissionController.grantPermissions.bind(
+          this.permissionController,
+        ),
+        findNetworkClientIdByChainId:
+          this.networkController.findNetworkClientIdByChainId.bind(
+            this.networkController,
+          ),
+        listAccounts: this.accountsController.listAccounts.bind(
+          this.accountsController,
+        ),
+        addNetwork: this.networkController.addNetwork.bind(
+          this.networkController,
+        ),
+        removeNetwork: this.networkController.removeNetwork.bind(
+          this.networkController,
+        ),
+        requestPermissionApprovalForOrigin: this.requestPermissionApproval.bind(
+          this,
+          origin,
+        ),
+        sendMetrics: this.metaMetricsController.trackEvent.bind(
+          this.metaMetricsController,
+        ),
+        metamaskState: this.getState(),
+        getCaveat: this.permissionController.getCaveat.bind(
+          this.permissionController,
+        ),
+        getSelectedNetworkClientId: () =>
+          this.networkController.state.selectedNetworkClientId,
+        revokePermission: this.permissionController.revokePermission.bind(
+          this.permissionController,
+        ),
+      }),
+    );
+
+    // TODO: Does this need to go before the wallet_createSession middleware?
+    // Add a middleware that will switch chain on each request (as needed)
+    const requestQueueMiddleware = createQueuedRequestMiddleware({
+      enqueueRequest: this.queuedRequestController.enqueueRequest.bind(
+        this.queuedRequestController,
+      ),
+      useRequestQueue: this.preferencesController.getUseRequestQueue.bind(
+        this.preferencesController,
+      ),
+      shouldEnqueueRequest: (request) => {
+        return methodsRequiringNetworkSwitch.includes(request.method);
+      },
+    });
+    engine.push(requestQueueMiddleware);
+
+    engine.push(
+      createUnsupportedMethodMiddleware([
+        ...UNSUPPORTED_RPC_METHODS,
+        'eth_requestAccounts',
+        'eth_accounts',
+      ]),
+    );
+
+    engine.push(
+      createMultichainMethodMiddleware({
+        subjectType: SubjectType.Website, // TODO: this should probably be passed in
```

Comment:
Definitely.

Reply:
done as part of this https://github.com/MetaMask/metamask-extension/pull/27782/commits/7e5782b74754cb188880a1230aaf220739fee1c3